File: 2017/day10.py
from pathlib import Path
from functools import reduce


def main():
    lengths = [int(length) for length in Path(
        "inputs/day10.txt").read_text().split(',')]
    nums = get_sparse_hash(lengths, list(range(256)), 1)
    print(f"Part 1: {nums[0] * nums[1]}")

    lengths = [ord(c) for c in list(Path("inputs/day10.txt").read_text())]
    lengths += [17, 31, 73, 47, 23]
    nums = get_sparse_hash(lengths, list(range(256)), 64)
    dense_hash = get_dense_hash(nums)
    print(f"Part 2: {list_to_hex_str(dense_hash)}")


def get_sparse_hash(lengths: list[int], nums: list[int], iters: int) -> list[int]:
    nums_len = len(nums)
    idx = 0
    skip_size = 0

    for i in range(iters):
        for length in lengths:
            start = idx
            end = idx + length - 1
            while start < end:
                a = start % nums_len
                b = end % nums_len
                nums[a], nums[b] = nums[b], nums[a]
                start += 1
                end -= 1
            idx = (idx + length + skip_size) % nums_len
            skip_size += 1
    return nums


def get_dense_hash(sparse_hash: list[int]) -> list[int]:
    return [reduce(lambda acc, curr: acc ^ curr, sparse_hash[i * 16: (i + 1) * 16]) for i in range(16)]


def list_to_hex_str(nums: list[int]) -> str:
    return ''.join([to_hex(num) for num in nums])

    
def to_hex(num: int) -> str:
    return hex(num)[2:].rjust(2, "0")

    

if __name__ == "__main__":
    main()

File: 2015/day10.js
import fs from 'fs';

let num = fs.readFileSync('inputs/day10.txt', 'utf8');

// pt. 1, i < 40
// pt. 2, i < 50
for (let i = 0; i < 50; i++) {
  let newNum = '';
  let lastChar = num[0];
  let count = 1;
  for (let j = 1; j < num.length; j++) {
    if (num[j] === lastChar) {
      count++;
    } else {
      newNum += count;
      newNum += lastChar;
      lastChar = num[j];
      count = 1;
    }
  }
  newNum += count;
  newNum += lastChar;
  num = newNum;
}

console.log(`Answer: ${num.length}`);